### ISSUE TITLE
Rules should have a clean return API

### ISSUE THREAD

**Issue by digitalkaoz** [phmLabs/Smoke#4]
they return strings (for failures) or booleans (for success), thats not very clear, think of throwing exceptions in case of a failure?

**Comment by phmLabs** [comment]
Or just return a Result object with getStatus(): Result::SUCCESS, Result::FAILURE

**Comment by digitalkaoz** [comment]
But you need the failure message?!

Nils Langner <notifications@github.com> schrieb am Do., 21. Mai 2015 09:55:

**Comment by phmLabs** [comment]
Sure, but there could be two methods getStatus() and getMessage()

**Comment by digitalkaoz** [comment]
why not simply throw? ;)

**Comment by phmLabs** [comment]
Exception it is.

**Comment by tomkyle** [comment]
Thx, useful. Proposal: As I understand, rules test on certain values given (i.e. file size or closing tags). If not matching, Exception is thrown. Thus, we should use generic SPL UnexpectedValueException class, which extends RuntimeException (and thus Exception). This would not cause BC break, would it?

**Comment by phmLabs** [comment]
This is a release candidate we can break whatever we want. What kind of exception does phpUnit use if an assertion failed?

**Comment by tomkyle** [comment]
As the PHPUNit's framework repo reads to me, all exceptions thrown inherit like this:

```
PHPUnit_Framework_ExpectationFailedException
extends PHPUnit_Framework_AssertionFailedError
extends PHPUnit_Framework_Exception
extends RuntimeException
```

TLDR Docblocks say:

- PHPUnit_Framework_ExpectationFailedException: "for expectations which failed their check"
- PHPUnit_Framework_AssertionFailedError: "when an assertion failed"
- PHPUnit_Framework_Exception: "Base class for all PHPUnit Framework exceptions"
- RuntimeException
